File: golang/wns/resolver.go
package wns

import (
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// ResolveHandle resolves a handle over HTTPS.
func ResolveHandle(ctx context.Context, handle string) (HandleResolutionDocument, error) {
	return ResolveHandleWithOptions(ctx, handle, ResolveHandleOptions{})
}

// ResolveHandleWithOptions resolves a handle with explicit options.
func ResolveHandleWithOptions(ctx context.Context, handle string, options ResolveHandleOptions) (HandleResolutionDocument, error) {
	bareHandle := strings.TrimPrefix(handle, "wba://")
	localPart, domain, err := ValidateHandle(bareHandle)
	if err != nil {
		return HandleResolutionDocument{}, err
	}
	normalized := localPart + "." + domain
	resolutionURL := BuildResolutionURL(localPart, domain)
	if options.BaseURLOverride != "" {
		resolutionURL = strings.TrimRight(options.BaseURLOverride, "/") + "/.well-known/handle/" + localPart
	}
	client := newWNSHTTPClient(options)
	request, err := http.NewRequestWithContext(ctx, http.MethodGet, resolutionURL, nil)
	if err != nil {
		return HandleResolutionDocument{}, err
	}
	request.Header.Set("Accept", "application/json")
	response, err := client.Do(request)
	if err != nil {
		return HandleResolutionDocument{}, fmt.Errorf("network error resolving handle '%s': %v", normalized, err)
	}
	defer response.Body.Close()
	if response.StatusCode == http.StatusMovedPermanently {
		return HandleResolutionDocument{}, fmt.Errorf("handle '%s' has been migrated", normalized)
	}
	if response.StatusCode == http.StatusNotFound {
		return HandleResolutionDocument{}, fmt.Errorf("handle '%s' does not exist", normalized)
	}
	if response.StatusCode == http.StatusGone {
		return HandleResolutionDocument{}, fmt.Errorf("handle '%s' has been permanently revoked", normalized)
	}
	if response.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(response.Body)
		return HandleResolutionDocument{}, fmt.Errorf("unexpected status %d resolving '%s': %s", response.StatusCode, normalized, string(body))
	}
	var document HandleResolutionDocument
	if err := authenticationDecodeJSON(response.Body, &document); err != nil {
		return HandleResolutionDocument{}, err
	}
	if strings.ToLower(document.Handle) != normalized {
		return HandleResolutionDocument{}, fmt.Errorf("handle mismatch: requested '%s', got '%s'", normalized, document.Handle)
	}
	dropInvalidProfileProjection(&document)
	return document, nil
}

// ResolveHandleFromURI resolves a wba:// URI.
func ResolveHandleFromURI(ctx context.Context, uri string) (HandleResolutionDocument, error) {
	parsed, err := ParseWBAURI(uri)
	if err != nil {
		return HandleResolutionDocument{}, err
	}
	return ResolveHandle(ctx, parsed.Handle)
}

func newWNSHTTPClient(options ResolveHandleOptions) *http.Client {
	timeout := 10 * time.Second
	if options.TimeoutSeconds > 0 {
		timeout = time.Duration(options.TimeoutSeconds * float64(time.Second))
	}
	transport := http.DefaultTransport.(*http.Transport).Clone()
	if options.VerifySSL != nil && !*options.VerifySSL {
		transport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
	}
	return &http.Client{Timeout: timeout, Transport: transport}
}

func authenticationDecodeJSON(reader io.Reader, target any) error {
	decoder := json.NewDecoder(reader)
	decoder.UseNumber()
	return decoder.Decode(target)
}

func dropInvalidProfileProjection(document *HandleResolutionDocument) {
	if document.Profile == nil {
		return
	}
	if !isKnownSubjectType(document.Profile.SubjectType) {
		document.Profile.SubjectType = SubjectTypeUnknown
	}
	if document.Profile.SubjectDID != document.DID {
		document.Profile = nil
		return
	}
	if document.Profile.Handle != "" && document.Profile.Handle != document.Handle {
		document.Profile = nil
	}
}

func isKnownSubjectType(value SubjectType) bool {
	switch value {
	case SubjectTypePerson,
		SubjectTypeAgent,
		SubjectTypeGroup,
		SubjectTypeOrganization,
		SubjectTypeService,
		SubjectTypeApplication,
		SubjectTypeUnknown:
		return true
	default:
		return false
	}
}

File: dart/lib/src/wns/resolver.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../errors.dart';
import 'types.dart';
import 'validator.dart';

Future<HandleResolutionDocument> resolveHandle(
  String handle, {
  ResolveHandleOptions options = const ResolveHandleOptions(),
  http.Client? client,
}) async {
  final parsed = validateHandle(handle.replaceFirst(RegExp(r'^wba://'), ''));
  final normalized = parsed.handle;
  final url = options.baseUrlOverride == null
      ? buildResolutionUrl(parsed.localPart, parsed.domain)
      : '${options.baseUrlOverride!.replaceFirst(RegExp(r'/+$'), '')}/.well-known/handle/${parsed.localPart}';
  final owned = client == null;
  final httpClient = client ?? http.Client();
  try {
    final response = await httpClient
        .get(Uri.parse(url), headers: const {'Accept': 'application/json'})
        .timeout(options.timeout);
    switch (response.statusCode) {
      case 200:
        break;
      case 301:
        throw AnpWnsException("handle '$normalized' has been migrated");
      case 404:
        throw AnpWnsException("handle '$normalized' does not exist");
      case 410:
        throw AnpWnsException(
          "handle '$normalized' has been permanently revoked",
        );
      default:
        throw AnpWnsException(
          "unexpected status ${response.statusCode} resolving '$normalized': ${response.body}",
        );
    }
    final decoded = jsonDecode(response.body);
    if (decoded is! Map) {
      throw const AnpWnsException('resolution document is not an object');
    }
    final map = Map<String, Object?>.from(decoded.cast<String, Object?>());
    final responseHandle =
        map['handle']?.toString().toLowerCase() ?? normalized;
    if (responseHandle != normalized) {
      throw AnpWnsException(
        "handle mismatch: requested '$normalized', got '$responseHandle'",
      );
    }
    final did = map['did']?.toString() ?? '';
    var profile = map['profile'] is Map
        ? DidSubjectProfile.fromJson(
            Map<String, Object?>.from(
              (map['profile'] as Map).cast<String, Object?>(),
            ),
          )
        : null;
    if (profile != null && profile.subjectDid != did) {
      profile = null;
    } else if (profile?.handle != null && profile!.handle != responseHandle) {
      profile = null;
    }
    return HandleResolutionDocument(
      handle: responseHandle,
      did: did,
      status: _parseHandleStatus(map['status']),
      updated: map['updated']?.toString(),
      versionId: map['versionId']?.toString(),
      ttl: map['ttl'] is int ? map['ttl'] as int : null,
      profile: profile,
      didDocument: map['didDocument'] is Map
          ? Map<String, Object?>.from(
              (map['didDocument'] as Map).cast<String, Object?>(),
            )
          : null,
    );
  } finally {
    if (owned) httpClient.close();
  }
}

Future<HandleResolutionDocument> resolveHandleFromUri(
  String uri, {
  ResolveHandleOptions options = const ResolveHandleOptions(),
  http.Client? client,
}) => resolveHandle(parseWbaUri(uri).handle, options: options, client: client);

HandleStatus _parseHandleStatus(Object? value) {
  final normalized = value?.toString().toLowerCase() ?? 'active';
  for (final status in HandleStatus.values) {
    if (status.name == normalized) return status;
  }
  throw AnpWnsException("unexpected handle status '$normalized'");
}
